### DISCUSSION TITLE
How to ensure calculated fields are requested using LoadItemsByCamlQuery>

### DISCUSSION THREAD

**Discussion by wizofaus** [pnp/pnpcore#1378] [Q&A]
The following does nothing (the expression is ignored):

`list.LoadItemsByCamlQuery("<View></View>", i => i["LinkFilename"]);`

Whereas this throws a 'System.NullReferenceException':

`list.LoadItemsByCamlQuery("<View></View>", i => i.QueryProperties(i => i["LinkFilename"]);`

But if the field is not explicitly requested, no value for it is returned.  This works fine using CSOM (Microsoft.SharePoint.Client library) using the "Include" syntax.

I tried "faking it" by creating a MemberExpression based on an explicitly defined property "LinkFilename" but couldn't get that to work either (the functions for creating Expressions in .NET core do a lot of paranoid type checking, not sure why!).  The only thing that actually seemed to (partially) work was to modify the PnP.Core source code to add "LinkFilename" as an actual property of IListItem/ListItem, but obviously that's not workable, as there's no fixed set of possible calculated fields for any SP list.

The other thing that works  (but requires access to PnP.Core internals  - I did this in the unit test project):
```
                await myList.LoadItemsByCamlQueryBatchAsync("<View></View>");
                var apiCall = context.CurrentBatch.Requests.First().Value.ApiCall;
                apiCall.Request += "?$select=Id,LinkFilename";
                context.CurrentBatch.Requests.First().Value.ApiCall = apiCall;
                await context.ExecuteAsync();
```
If you don't have access to the internal "ApiCall" property it has to be run via reflection.

I also tried:

```
                var apiRequest = new ApiRequest(System.Net.Http.HttpMethod.Post, ApiRequestType.SPORest, 
                    $"_api/web/Lists(guid'{myList.Id}')/GetItems?$select=Id,LinkFilename",
                    "{\"query\":{}}");
                var response = await context.Web.ExecuteRequestAsync(apiRequest);
```
But sadly the response is not mapped back to the Items model, so rather defeats the purpose of using the PnP.Core library in the first place (again, I could use ExecuteRequestBatchAsync, and reset the "RawRequest" flag of the ApiCall via reflection, but I'm trying to find a support way of doing this that won't break with future implementation detail changes!).

So how can I explicitly request that this field is loaded?

**Comment by jansenbe** [comment]
@wizofaus : either loading the calculated field by specifying it in the CAML query or by loading via `FieldValuesAsText` works in my test.

```csharp
// Query the document again
const string viewXml = @"<View Scope='Recursive'>
                            <ViewFields>
                                <FieldRef Name='Title' />
                                <FieldRef Name='Description' />
                                <FieldRef Name='calc' />
                            </ViewFields>
                            <RowLimit>1</RowLimit>
                        </View>";

var list = await context.Web.Lists.GetByTitleAsync("calculatedfields");
Expression<Func<IListItem, object>>[] selectors =
{
    li => li.All,
    li => li.FieldValuesAsText
};

await list.LoadItemsByCamlQueryAsync(new CamlQueryOptions()
{
    ViewXml = viewXml,
    DatesInUtc = true
}, selectors).ConfigureAwait(false);

Assert.IsTrue(list.Items.AsRequested().First()["Title"].ToString() == "Bert");
Assert.IsTrue(list.Items.AsRequested().First()["Description"].ToString() == "Jansen");
Assert.IsTrue(list.Items.AsRequested().First()["calc"].ToString() == "BertJansen");

Assert.IsTrue(list.Items.AsRequested().First().FieldValuesAsText["calc"].ToString() == "BertJansen");
```

**Comment by wizofaus** [comment]
FieldValuesAsText definitely didn't give me the values I was looking for when I tried. But still not sure if I need this functionality just yet. Just seemed odd there was no way to add the query parameters that the API can take.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by jansenbe** [reply]
@wizofaus : this is because we pick up fields defined in the CAML query to update the issued API call, see https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Internal/List.cs#L456-L493. That explains why you did not see the field getting populated in FieldValuesAsText . When you add it to the CAML query is will be included.
